ls1intum/Artemis
# `Programming exercises`: Allow users to configure the tab width in the online editor

### Checklist
#### General


#### Client

### Motivation and Context
Changing the tab size is both important for students and tutors:
- Some programming languages do not use four spaces as default. Students should be able to easily use e.g., two spaces instead in the online editor.
- Some students might use proper TAB characters in their local editor. There it might be displayed in a width of eight. When the tutor then looks at the code in the online editor with a fixed tab size of four the formatting is pretty broken. By allowing the tutor to change how tabs are displayed even in read-only mode, the formatting as intended by the student can be displayed.

### Description
The code editor already has an option to set how tabs are displayed/how many spaces are inserted when pressing the TAB key in edit mode. This PR makes this setting available to the user.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise

1. Submit to the programming exercise using the online editor. When changing the tab-width the editor should insert the corresponding number of spaces instead.
2. Checkout the repository locally and submit a file that contains actual TAB characters.
3. Open the submission as the instructor. Changing the tab size should change the display for the file with the actual TAB characters.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| `code-editor-ace.component.ts` | 56% | 70.3% |
| `ace-editor.component.ts` | 68% | 84.4% |

all new code covered 2x 100%

### Screenshots
Example in read-only mode when assessing submissions
![Screenshot 2022-03-23 at 17-04-04 Programming Exercises](https://user-images.githubusercontent.com/64250573/159744235-2981174e-5445-46cd-89a4-b235d5eef997.png)
![Screenshot 2022-03-23 at 17-04-16 Programming Exercises](https://user-images.githubusercontent.com/64250573/159744237-40e70436-6f5d-4ede-8c74-cc14ccf6cb04.png)

Design of the settings menu changed since the screenshots above, effect of the change is still the same:
![Screenshot 2022-04-05 at 17-41-50 Code Editor](https://user-images.githubusercontent.com/64250573/161817390-9c43c5e4-af2f-475f-acd6-21ecb17f4810.png)
